### PR TITLE
posix: atime update improvements

### PR DESCRIPTION
**Pull request by marcinslusarz**
Builds on top of #393

---
This change is [Reviewable](https://reviewable.io/reviews/pmem/pmemfile/402)

### CODE REVIEW

File: src/libpmemfile-posix/read.c [hunk 1]
```diff
@@ -189,15 +189,14 @@ handle_atime(PMEMfilepool *pfp,
...
-	ASSERT_NOT_IN_TX();
-
 	os_rwlock_wrlock(&vinode->rwlock);
 
-	TX_BEGIN_CB(pfp->pop, cb_queue, pfp) {
-		inode_tx_set_atime(inode, tm);
-	} TX_ONABORT {
-		LOG(LINF, "can not update inode atime");
-	} TX_END
+	bool atime_slot = inode_next_atime_slot(inode);
```
Comment (GBuella) [inline src/libpmemfile-posix/read.c]: Didn't we have a typedef for this bool?

File: src/libpmemfile-posix/write.c [hunk 2]
```diff
@@ -301,34 +301,50 @@ pmemfile_pwritev_internal(PMEMfilepool *pfp,
 	 * only non-content related metadata changed, so it's safer to do it.
 	 */
 	bool update_mtime = (tm_diff >= 1000000) || update_size;
+	bool update_atime = vinode->atime_dirty &&
+				(update_mtime || update_size);
 
 	bool size_slot = inode->slots.bits.size;
 	bool ctime_slot = inode->slots.bits.ctime;
+	bool atime_slot = inode->slots.bits.atime;
```
Comment (GBuella) [inline src/libpmemfile-posix/write.c]: `inode_slot atime_slot`

File: src/libpmemfile-posix/read.c (hunk 1 above)
Comment (marcinslusarz) [reply src/libpmemfile-posix/read.c]: Fixed.

File: src/libpmemfile-posix/write.c (hunk 2 above)
Comment (marcinslusarz) [reply src/libpmemfile-posix/write.c]: Done.

**Comment by sarahjelinek** [comment]
Reviewed 15 of 34 files at r1, 16 of 19 files at r2.
Review status: 31 of 34 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/libpmemfile-posix/inode.c, line 811 at r2](https://reviewable.io:443/reviews/pmem/pmemfile/402#-Kvc1jqc0tSOQ2nqj_gO:-Kvc1jqdz9dWDKsBEVNn:bmn53vl) ([raw file](https://github.com/pmem/pmemfile/blob/74076a08842edb47a90a3bd9a154061fc9596ab3/src/libpmemfile-posix/inode.c#L811)):*
> ```C
> 	if (vinode->atime_dirty)
> 		/* we did it at suspend time */
> 		vinode->atime_dirty = false;
> ```

Why don't you clear the atime_dirty at suspend?

---

*[src/libpmemfile-posix/inode.h, line 242 at r2](https://reviewable.io:443/reviews/pmem/pmemfile/402#-Kvc21hwTaZmqyHmfFIQ:-Kvc21hwTaZmqyHmfFIR:bz24gnx) ([raw file](https://github.com/pmem/pmemfile/blob/74076a08842edb47a90a3bd9a154061fc9596ab3/src/libpmemfile-posix/inode.h#L242)):*
> ```objc
> 
> static inline void
> pmemfile_tx_time_set(struct pmemfile_time *time, struct pmemfile_time tm)
> ```

since the other functions are set_xxxtime() this one should be set_time()? Just for consistency? Unless you named this slightly different for a reason.

---

*[src/libpmemfile-posix/layout.h, line 145 at r2](https://reviewable.io:443/reviews/pmem/pmemfile/402#-Kvc3CZYXN3ur2Te54l6:-Kvc3CZYXN3ur2Te54l7:bxvot80) ([raw file](https://github.com/pmem/pmemfile/blob/74076a08842edb47a90a3bd9a154061fc9596ab3/src/libpmemfile-posix/layout.h#L145)):*
> ```objc
> #define PMEMFILE_INODE_SIZE METADATA_BLOCK_SIZE
> #define PMEMFILE_IN_INODE_STORAGE \
> 	(sizeof(struct pmemfile_dir) + 2 * sizeof(struct pmemfile_dirent) + 8)
> ```

I am not sure I understand why you made this change?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/402)*

**Comment by marcinslusarz** [comment]
Do you want me to rebase it to master to get rid of commits from #393 ?

---

Review status: 31 of 34 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/libpmemfile-posix/inode.c, line 811 at r2](https://reviewable.io:443/reviews/pmem/pmemfile/402#-Kvc1jqc0tSOQ2nqj_gO:-KvcBStibibROvf4W1ld:b-fwz467) ([raw file](https://github.com/pmem/pmemfile/blob/74076a08842edb47a90a3bd9a154061fc9596ab3/src/libpmemfile-posix/inode.c#L811)):*
<details><summary><i>Previously, sarahjelinek (Sarah Jelinek) wrote…</i></summary><blockquote>

Why don't you clear the atime_dirty at suspend?
</blockquote></details>

Because suspend is in transaction - if it fails it means atime was not written back to medium.

---

*[src/libpmemfile-posix/inode.h, line 242 at r2](https://reviewable.io:443/reviews/pmem/pmemfile/402#-Kvc21hwTaZmqyHmfFIQ:-KvcEHR3ANAYSxbLvdop:b-tagsms) ([raw file](https://github.com/pmem/pmemfile/blob/74076a08842edb47a90a3bd9a154061fc9596ab3/src/libpmemfile-posix/inode.h#L242)):*
<details><summary><i>Previously, sarahjelinek (Sarah Jelinek) wrote…</i></summary><blockquote>

since the other functions are set_xxxtime() this one should be set_time()? Just for consistency? Unless you named this slightly different for a reason.
</blockquote></details>

It's part of #393 this PR builds on. 

Other functions are named inode_tx_set_xtime because they operate on an inode, do it in transaction and they set xtime.

This function operates on pmemfile_time, in transaction and sets its value, so it should be either pmemfile_time_tx_set or time_tx_set. I added one commit to rename it to latter one.

---

*[src/libpmemfile-posix/layout.h, line 145 at r2](https://reviewable.io:443/reviews/pmem/pmemfile/402#-Kvc3CZYXN3ur2Te54l6:-KvcCsdD5JL7N1wRKm-O:b-celmrq) ([raw file](https://github.com/pmem/pmemfile/blob/74076a08842edb47a90a3bd9a154061fc9596ab3/src/libpmemfile-posix/layout.h#L145)):*
<details><summary><i>Previously, sarahjelinek (Sarah Jelinek) wrote…</i></summary><blockquote>

I am not sure I understand why you made this change?
</blockquote></details>

It's part of #393 (specifically commit 953ebd3efeea261c576544bdd6c94a5c3e308394). 

I did it because I aligned the beginning of pmemfile_inode.file_data to cacheline boundary and wanted to use the remaining 8 bytes hole for short symlink information.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/402#-:-KvcKEtQjyIPWMUB1uzR:ba406e)*

**Comment by sarahjelinek** [comment]
Yes. thanks.

---

Review status: 31 of 34 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/402#-:-KvcL58i0xKpvl7agrCK:b-8x5427)*

**Comment by GBuella** [comment]
Reviewed 1 of 19 files at r2.
Review status: 32 of 34 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/402)*

**Comment by sarahjelinek** [comment]
:lgtm:

---

Review status: 5 of 6 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/402#-:-KvcNYeZPpjQXWq8mBIW:bnfp4nl)*

**Pull request merged**